### DISCUSSION TITLE
Raw queries

### DISCUSSION THREAD

**Discussion by iorlas** [go-pg/pg#1876] [Q&A] [answered]
Just a quick question/suggestion. Well, I have an app which uses go-pg mainly. It looks nice and sound, but I have to do some raw queries with variable results. The use-case could be described like this: "advanced version of raw SQL API" (hasura or any existing BI or ETL solution did not fit the needs damn). 

So, input is predictable in one way or another, the query will be quite complex (CTEs, window-functions, sub-queries, complex joins). The output (columns) is kind of predictable, but not really.

In general, it is easy to implement with low-level drivers by making the iteration and scanning each row into the map, interface{} or just a slice of strings. It will allow to stream the data, make less noise in the code, great flexibility.

It is easily doable with database/sql, but might not be that easy with go-pg. I would like to avoid keeping two separate database drivers, keeping 2 db pools. So, here are the pitfalls I found:

- Query.ForEach is the only way to iterate w/o scanning all the data into memory. But it needs query object. I need to supply raw query, but both Exec and Query return Result instead, making ForEach non-applicable
- I need to unmarshal a variable num of columns, so I would expect to pass `[][]string` as a model. Obviously, it does not work. I've tried to use `pg.Strings`, but it did not work either (the issue with unmarshalling, it might be my mistake somewhere, will take a look on weekends, too overwhelmed).

Well, not a requirement, just another usecase, it would be awesome to have some flexibility. I believe go-pg is quite flexible right now, but it might not be as obvious as should be.

Anyway, wanted to give you some feedback asap, hope it helps you, might help me if you will suggest the preferred approach :)

*Btw, we redid the otel tracing hook almost from the zero, to be compliant with the specs, will send you PR on weekends as well. There are a couple of issues and some optimizations still to be done, but the current otel tracing hook has some critical issues and missing essential data, we'll see*

**Comment by elliotcourant** [comment]
> Query.ForEach is the only way to iterate w/o scanning all the data into memory. But it needs query object. I need to supply raw query, but both Exec and Query return Result instead, making ForEach non-applicable

Not to say that this is the best solution, but it might be _a_ solution. What about making a model for that query and then making the query a view in SQL? Then go-pg would simply select from that view as if it were a table allowing you to use your `ForEach` normally.

**Reply by iorlas** [reply]
To create a model and a view, we need to know the column names beforehand. But the point is to provide pool-ppl-BI to make some simple queries with different columns.

It is an option to convert the columns into json and expect that json column in the model, but yeah.. painful.

**Comment by vmihailenco** [comment, marked as answer]
Questions like this is one of the reasons I've created [Bun](https://bun.uptrace.dev/). It works on top of database/sql so `sql.Rows` is always available.

**Reply by iorlas** [reply]
It works, kinda.. If the answer is to use another software it is not really a solution within the product itself. TBH it is really inconvenient to have multiple PG drivers at the same time. I believe go-pg should have such low-level API, but it is kind of a fundamental issue, so I guess it is unlikely to see it anytime soon...

**Reply by iorlas** [reply]
`Bun is a rewrite of go-pg that works on top of sql.DB instead of using custom API`

A-ha! So, I guess, sometime in the future, we might see go-pg been fully rewritten on top of the Bun, am I right? :)

**Reply by vmihailenco** [reply]
>I believe go-pg should have such low-level API

go-pg with such low-level API is called Bun. I could release it as pg v11 instead. It would not change much.

**Reply by iorlas** [reply]
Theoretically speaking it is, indeed. But we cannot use GO-PG, install Bun, and pass GO-PG connection pool to the Bun, am I right?

We will need to create another connection pool with database/sql and manage it separately from the GO-PG, right?

This becomes a hell because typically we have some limits on num of connections. Like, 500. Then we assume our pool is limited to the 10 connections, we can use 50 instances. But if we will use 2 connection pools, we are stuck with an issue:

- If we use 5 in GO-PG and 5 in Bun, we lose efficiency
- If we use 10 in GO-PG and 10 in Bun, we lose scalability & efficiency
- Implement the shared limit? Well. could help, but we still lose some efficiency when connections cannot be reused by another pool.
- Tune to the project needs? Well, we already do that, 2 pools per instance are squaring the scalability problem.

**Reply by vmihailenco** [reply]
It would be the same for pg v10 and pg v11.

**Reply by iorlas** [reply]
Diving a little bit deeper into the docs I realized the Bun is a replacement for the go-pg with the intent to deprecate the go-pg. Am I right?

I guess we will evaluate the Bun as a migration step to use the same connection pool for raw and ORM load. Initially, I thought the Bun is just a simple query builder, but now it looks much better! Thank you for your efforts!

Sorry for the offtopic, but do you plan to deprecate go-pg in favor of the Bun? Any timelines? Does it make sense to invest any time in go-pg right now or you would recommend starting with Bun right away?

**Reply by vmihailenco** [reply]
>Diving a little bit deeper into the docs I realized the Bun is a replacement for the go-pg with the intent to deprecate the go-pg. Am I right?

Yes.

Practically nothing drastic happened. Bun is just pg v11 with a different core based on database/sql interfaces. pg v10 will only receive security updates and important bug fixes. The same happened with pg v9.

New projects should use Bun.

**Reply by iorlas** [reply]
Sweet! I suggest highlighting the project status a bit more, with big letters and red colors, you know. I see this info in the readme, but one of our project just started to use go-pg, feels like this info is easily missed.

Again, sorry for the offtopic, but do you plan to provide the soft-delete capability in the Bun? I didn't find it in the models or deletion docs. We use it greatly in our projects, it might not be as important as other stuff, but it would be awesome to know the status.

**Reply by vmihailenco** [reply]
Yes, soft deletes are supported - https://github.com/uptrace/bun/blob/master/internal/dbtest/soft_delete_test.go#L56-L87

>with big letters and red colors

I will add something like that when Bun reaches v1.0 (July or August I guess).

**Reply by iorlas** [reply]
Aah, didn't spot it right away. I guess the docs have some gaps, but it is fine for now. Thank you for your work!

To the next ppl looking into the issue: migrate off the go-pg to the Bun, which could be used as a ORM and Raw SQL executer since it is built on top of the database/sql package.
